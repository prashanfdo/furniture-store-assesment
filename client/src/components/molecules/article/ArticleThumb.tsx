import styled from "@emotion/styled";
import { Article } from "types";
import AddToCartButton from "./AddToCartButton";
import ArticleName from "./ArticleName";
import ArticleImages from "./ArticleImages";

const LinkContainer = styled.a`
  display: block;
  border: 0;
  border-radius: 10px;
  background-color: #fff;
  overflow: hidden;
  text-decoration: none;
  padding: 0.5rem;
`;

type ArticleThumbProps = {
  product: Article;
};
const ArticleThumb: React.FC<ArticleThumbProps> = ({ product }: ArticleThumbProps) => {
  return (
    <LinkContainer href="#product-page">
      <ArticleImages productImages={product.images} />
      <ArticleName product={product} />
      <AddToCartButton product={product} />
    </LinkContainer>
  );
};

export default ArticleThumb;
