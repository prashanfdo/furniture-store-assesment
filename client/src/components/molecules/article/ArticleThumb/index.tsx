import { Link } from "react-router-dom";
import { Article } from "types";
import { AddToCartButton, ArticleImage, ArticleName, ArticlePrice } from "components/atoms";

type ArticleThumbProps = {
  product: Article;
  // TODO: make this required once add to cart feature implemented
  onAddtoCart?: (product: Article) => void;
};
const ArticleThumb: React.FC<ArticleThumbProps> = ({ product, onAddtoCart }: ArticleThumbProps) => {
  const handleAddToCartButtonClick = () => {
    onAddtoCart?.(product);
  };

  return (
    <Link
      className="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      to="#product-page"
    >
      <ArticleImage src={product.images[0].path} alt={product.name} />
      <ArticleName>{product.name}</ArticleName>
      <ArticlePrice
        currency={product.prices.currency}
        price={product.prices.regular.value}
        salePrice={product.prices.special?.value}
        discount={product.prices.special?.discount}
      />
      <AddToCartButton onClick={handleAddToCartButtonClick} />
    </Link>
  );
};

export default ArticleThumb;
