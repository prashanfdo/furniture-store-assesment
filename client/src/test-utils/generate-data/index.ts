export * from "./article";
export * from "./category";
export * from "./search";
