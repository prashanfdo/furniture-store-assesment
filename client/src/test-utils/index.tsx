import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
