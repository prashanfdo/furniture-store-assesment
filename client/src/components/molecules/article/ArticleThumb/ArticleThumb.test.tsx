import { createArticle, render, screen, user } from "test-utils";
import ArticleThumb from "./index";

const dataProduct: React.ComponentProps<typeof ArticleThumb>["product"] = createArticle(true);

describe("ArticleThumb", () => {
  beforeEach(() => {
    user.setup();
  });

  it("should render", () => {
    const handleAddToCart = jest.fn();
    const { container } = render(<ArticleThumb product={dataProduct} onAddtoCart={handleAddToCart} />);
    expect(container).toMatchSnapshot();
  });

  it("should render name, image", async () => {
    const handleAddToCart = jest.fn();
    const { container } = render(<ArticleThumb product={dataProduct} onAddtoCart={handleAddToCart} />);
    expect(container.innerHTML).toContain(dataProduct.name);
    expect(container.innerHTML).toContain(dataProduct.images[0].path);
  });

  it("should invoke onAddtoCart callback", async () => {
    const handleAddToCart = jest.fn();
    render(<ArticleThumb product={dataProduct} onAddtoCart={handleAddToCart} />);
    await user.click(screen.getByRole("button"));
    expect(handleAddToCart).toHaveBeenCalledTimes(1);
  });
});
