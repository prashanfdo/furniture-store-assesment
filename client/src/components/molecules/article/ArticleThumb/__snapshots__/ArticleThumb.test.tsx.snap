// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleThumb should render 1`] = `
<div>
  <a
    class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
    href="/#product-page"
  >
    <div>
      <img
        alt="Modern Frozen Fish"
        class="object-contain w-full h-96"
        loading="lazy"
        src="https://loremflickr.com/640/480"
      />
    </div>
    <h4
      class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
    >
      Modern Frozen Fish
    </h4>
    <div
      class="text-gray-600"
    >
      <span
        class="font-semibold text-red-600"
        data-testid="final-price"
      >
        785,00 €
      </span>
      <span
        class="ml-4 text-sm line-through"
        data-testid="regular-price"
      >
        1.000,00 €
      </span>
      <span
        class="px-1 py-1 ml-2 text-sm text-white bg-red-500 rounded-lg"
        data-testid="discount-percent"
      >
        -
        22
        %
      </span>
    </div>
    <button
      class="block w-full p-2 mt-2 text-center text-gray-600 bg-white border border-solid rounded cursor-pointer border-coral-200 hover:border-coral-600 hover:bg-coral-600 hover:text-white"
    >
      Add To Cart
    </button>
  </a>
</div>
`;
